superset-frontend/src/SqlLab/components/QueryTable/index.tsx

```diff
@@ -29,133 +28,166 @@ import Button from 'src/components/Button';
 
-    const openQueryInNewTab = query => {
-      props.actions.cloneQueryToNewTab(query, true);
+    const openQueryInNewTab = (query: Query) => {
+      actions?.cloneQueryToNewTab(query, true);
     };
 
-    const openAsyncResults = (query, displayLimit) => {
-      props.actions.fetchQueryResults(query, displayLimit);
+    const openAsyncResults = (query: Query, displayLimit: number) => {
+      actions?.fetchQueryResults(query, displayLimit);
     };
 
-    const clearQueryResults = query => {
-      props.actions.clearQueryResults(query);
+    const clearQueryResults = (query: Query) => {
+      actions?.clearQueryResults(query);
     };
 
-    const removeQuery = query => {
-      props.actions.removeQuery(query);
+    const removeQuery = (query: Query) => {
+      actions?.removeQuery(query);
     };
 
-    return props.queries
+    const statusAttributes = {
+      success: {
+        config: {
+          icon: <Icons.Check iconColor={theme.colors.success.base} />,
+          label: t('Success'),
+        },
+      },
+      failed: {
+        config: {
+          icon: <Icons.XSmall iconColor={theme.colors.error.base} />,
+          label: t('Failed'),
+        },
+      },
+      stopped: {
+        config: {
+          icon: <Icons.XSmall iconColor={theme.colors.error.base} />,
+          label: t('Failed'),
+        },
+      },
+      running: {
+        config: {
+          icon: <Icons.Running iconColor={theme.colors.primary.base} />,
+          label: t('Running'),
+        },
+      },
+      fetching: {
+        config: {
+          icon: <Icons.Queued iconColor={theme.colors.primary.base} />,
+          label: t('fetching'),
+        },
+      },
+      timed_out: {
+        config: {
+          icon: <Icons.Offline iconColor={theme.colors.grayscale.light1} />,
+          label: t('Offline'),
+        },
+      },
+      scheduled: {
+        config: {
+          icon: <Icons.Queued iconColor={theme.colors.grayscale.base} />,
+          label: t('Scheduled'),
+        },
+      },
+      pending: {
+        config: {
+          icon: <Icons.Queued iconColor={theme.colors.grayscale.base} />,
+          label: t('Scheduled'),
+        },
+      },
+      error: {
+        config: {
+          icon: <Icons.Error iconColor={theme.colors.error.base} />,
+          label: t('Unknown Status'),
+        },
+      },
+    };
+
+    return queries
       .map(query => {
-        const q = { ...query };
-        const status = statusAttributes[q.state] || statusAttributes.error;
+        // query's type is original Query; Shallow-copy of query, q's type is QueryTableQuery. So that prop, sql passed to another component will remain string, the type of original Query.
+        // qTemp is used to bring in the properties that you don't overwrite, which will be assigned to q, so that q.state/q.sql/q.progress will be later object.
+        const { state, sql, progress, ...qTemp } = query;
+        const q: QueryTableQuery = { ...qTemp };
```

Comment:
do you need this extra const line, or can you just combine it with the previous line, like
`const { state, sql, progress, ...q } = query;`

Reply:
I guess I would need qTemp, bc only having { state, sql, progress, ...q } = query, I cannot add extra property/ key-value-pair to it, such as q.sql/state/progress as object. Please let me know if my understanding is not correct or there is other way to do it. Thank you very much.

Reply:
It should work out to be the same thing. with `const { state, sql, progress, ...q } = query;` you're creating an object called q which is a copy of query, minus state, sql and progress. 

in 
```
        const { state, sql, progress, ...qTemp } = query;
        const q: QueryTableQuery = { ...qTemp };
```
you create a copy of query called qTemp without state, sql, and progress, and then make another copy of it and call it q.

Reply:
Please correct me if my understanding is not correct. Thank you in advance. Bc QueryTableQuery is assigned as type to q, with state/sql/progress as object, then later i can add those extra key-value pair. If only { state, sql, progress, ...q } = query,  q has NO state/sql/progress, which will give me error to add state/sql/progress as object later.

Reply:
good question. You can assign types when you declare the variables in one line by writing:
`const { state, sql, progress, ...q } : {state: string; sql: string; q: QueryTableQuery} = query;`

Reply:
By using the above, i got an error =>> Property 'q' is missing in type 'Query' but required in type '{ state: string; sql: string; q: QueryTableQuery; }'.   I cannot find a way to fix it, as I guess query's type is Query. May I please have your suggestion here? Thank you!

Reply:
OK, let's keep what you have mostly, and just tweak it a bit to read a little more clear what's happening. 
Here's what I suggest: 
```
        const { state, sql, progress, ...rest } = query;
        const q = rest as QueryTableQuery;
 ```
This way we can clearly see that we're taking the rest of the properties and assigning them to a new variable which is remapping the type from one to another. LMK if that works. And thank you for your patience on this!

Reply:
Thank you. Updated as suggested. Learnt a lot from your comments. Really appreciate that.

Reply:
Thank you for all your incredible work on this PR!